homeassistant/components/bluetooth/active_update_coordinator.py

```diff
@@ -0,0 +1,122 @@
+from collections.abc import Callable, Coroutine
+import logging
+import time
+from typing import Any, TypeVar
+
+from homeassistant.core import HomeAssistant, callback
+
+from . import BluetoothChange, BluetoothScanningMode, BluetoothServiceInfoBleak
+from .passive_update_processor import (
+    PassiveBluetoothDataUpdate,
+    PassiveBluetoothProcessorCoordinator,
+)
+
+_T = TypeVar("_T")
+
+
+class ActiveBluetoothProcessorCoordinator(PassiveBluetoothProcessorCoordinator):
+    """
+    A coordinator that parses passive data from advertisements but can also poll.
+
+    Every time an advertisement is received, needs_poll_method is called to work
+    out if a poll is needed. This should return True if it is and False if it is
+    note.
+
+    def needs_poll_method(svc_info: BluetoothServiceInfoBleak, last_poll: float | None) -> bool:
+        return True
+
+    If there has been no poll since HA started, `last_poll` will be None. Otherwise it is
+    the number of seconds since one was last attempted.
+
+    If a poll is needed, the coordinator will call poll_method. This is a coroutine.
+    It should return the same type of data as your update_method. The expectation is that
+    data from advertisements and from polling are being parsed and fed into a shared
+    object that represents the current state of the device.
+
+    async def poll_method(svc_info: BluetoothServiceInfoBleak) -> YourDataType:
+        return YourDataType(....)
+
+    BluetoothServiceInfoBleak.device contains a BLEDevice. You should use this in
+    your poll function, as it is the most efficient way to get a BleakClient.
+    """
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        logger: logging.Logger,
+        address: str,
+        mode: BluetoothScanningMode,
+        update_method: Callable[[BluetoothServiceInfoBleak], _T],
+        needs_poll_method: Callable[[BluetoothServiceInfoBleak, float | None], bool],
+        poll_method: Callable[
+            [BluetoothServiceInfoBleak],
+            Coroutine[Any, Any, PassiveBluetoothDataUpdate[_T]],
+        ],
+    ) -> None:
+        """Initialize the processor."""
+        super().__init__(hass, logger, address, mode, update_method)
+
+        self._needs_poll_method = needs_poll_method
+        self._poll_method = poll_method
+        self._last_poll: float | None = None
+        self._last_poll_successful = True
+        self._poll_limiter = asyncio.Semaphore(2)
+
+    def needs_poll(self, service_info: BluetoothServiceInfoBleak) -> bool:
+        """Return true if time to try and poll."""
+        poll_age: float | None = None
+        if self._last_poll:
+            poll_age = time.monotonic() - self._last_poll
+        return self._needs_poll_method(service_info, poll_age)
+
+    async def async_poll(self, service_info: BluetoothServiceInfoBleak) -> None:
+        """Poll the device to retrieve any extra data."""
+        # There is already 1 poll running and 1 poll waiting
```

Comment:
What's the difference between running and waiting? We allow two acquisitions of the semaphore so two polls will be able to start, during the same time interval, before the first poll is finished.

Reply:
Ooops, the goal was to have never have more than one running, but that if a poll request came whilst the first was still in flight it could wait. But a second couldn't.

2 things though - (a) i screwed it up in a refactor and (b) it doesn't make sense - some of these devices send 4 or more advertisements at once. So you always double poll, and thats not what we want. Changed it to just a lock.

Reply:
This is resolved by using the debouncer.